griptape/tokenizers/base_tokenizer.py

```diff
@@ -1,50 +1,98 @@
 from __future__ import annotations
+from typing import Any
 from abc import ABC, abstractmethod
 from attrs import define, field, Factory
-from griptape import utils
+from griptape.utils import PromptStack
 
 
 @define()
 class BaseTokenizer(ABC):
+    DEFAULT_MAX_INPUT_TOKENS = 4096
+    DEFAULT_MAX_OUTPUT_TOKENS = 1000
     MODEL_PREFIXES_TO_MAX_INPUT_TOKENS = {}
     MODEL_PREFIXES_TO_MAX_OUTPUT_TOKENS = {}
 
     model: str = field(kw_only=True)
-    stop_sequences: list[str] = field(default=Factory(lambda: [utils.constants.RESPONSE_STOP_SEQUENCE]), kw_only=True)
-    max_input_tokens: int = field(kw_only=True, default=None)
-    max_output_tokens: int = field(kw_only=True, default=None)
+    stop_sequences: list[str] = field(default=Factory(list), kw_only=True)
+    max_input_tokens: int = field(
+        kw_only=True, default=Factory(lambda self: self._default_max_input_tokens(), takes_self=True)
+    )
+    max_output_tokens: int = field(
+        kw_only=True, default=Factory(lambda self: self._default_max_output_tokens(), takes_self=True)
+    )
 
-    def __attrs_post_init__(self) -> None:
-        if self.max_input_tokens is None:
-            self.max_input_tokens = self._default_max_input_tokens()
-
-        if self.max_output_tokens is None:
-            self.max_output_tokens = self._default_max_output_tokens()
-
-    def count_input_tokens_left(self, text: str | list) -> int:
+    def count_input_tokens_left(self, text: str | PromptStack) -> int:
         diff = self.max_input_tokens - self.count_tokens(text)
 
         if diff > 0:
             return diff
         else:
             return 0
 
-    def count_output_tokens_left(self, text: str | list) -> int:
+    def count_output_tokens_left(self, text: str | PromptStack) -> int:
         diff = self.max_output_tokens - self.count_tokens(text)
 
         if diff > 0:
             return diff
         else:
             return 0
 
+    def count_tokens(self, text: str | PromptStack) -> int:
+        if isinstance(text, PromptStack):
+            return self.try_count_tokens(self.prompt_stack_to_string(text))
+        else:
+            return self.try_count_tokens(text)
+
     @abstractmethod
-    def count_tokens(self, text: str | list[dict]) -> int: ...
+    def try_count_tokens(self, text: Any) -> int: ...
+
+    def prompt_stack_to_string(self, prompt_stack: PromptStack) -> str:
+        """Converts a Prompt Stack to a string for token counting or model input.
+        This base implementation will not be very accurate, and should be overridden by subclasses with model-specific tokens.
```

Comment:
Originally [same question](https://github.com/griptape-ai/griptape/pull/840/files#r1633879326) here, but I sense that the model-specific method in its own tokenizers means this should be owned by the tokenizer.

Reply:
Honestly we could probably get rid of this method. All of our Prompt Drivers now support message-style APIs.

Reply:
Ah but now I remember why I didn't remove in this PR: I didn't want to remove the token counts in the events which relies on this method. My next PR changes this functionality so I will remove this method.